### PR TITLE
Photo refactor

### PR DESCRIPTION
**Pull request by yuvipanda**
Moves photo upload functionality into a separate photo object. Stores only that in the database, rather than individual fields. 

Next step -> make the database photo.js aware, use same display for detail from 'my uploads' screen and the 'your upload has finished' screen and ponies.

### CODE REVIEW

File: assets/www/js/photo.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+	// Page content generated, monument data, user who took it
+	//
+	// Should also handle actual Uploading API, and adding stuff to the database
+	//
+	// Should add an entry right after it is 'created' (not via constructor)
+	// Should be able to mark it as uploaded when it is uploaded
+	//
+	// Make this generic enough to be reusable across projects
+	function Photo( data ) {
+		this.data = data;
```
Comment (jdlrobson) [inline assets/www/js/photo.js:17]: mm... wouldn't it be nicer for the photo to have properties such as fileTitle and fileContent ?

File: assets/www/js/database.js [hunk 2]
```diff
@@ -51,9 +51,10 @@ define( [ 'jquery' ], function( $ ) {
 		return d.promise();
 	}
 
-	function addUpload( monument, username, fileUrl, url, title, completed ) {
-		var insertSQL = "INSERT INTO uploads ( username, file, monument, url, title, timestamp, completed ) VALUES ( ?, ?, ?, ?, ?, ?, ? );";
-		execute( insertSQL, [ username, fileUrl, JSON.stringify( monument ), url, title, (new Date()).getTime() , completed ] );
+	function addUpload( username, monument, photo, completed ) {
+		var timestamp = (new Date()).getTime();
```
Comment (jdlrobson) [inline assets/www/js/database.js:55]: any reason for brackets around new Date() ?

File: assets/www/js/photo.js (hunk 1 above)
Comment (yuvipanda) [inline assets/www/js/photo.js:17]: True, I'll just move data onto the object itself.

File: assets/www/js/database.js (hunk 2 above)
Comment (yuvipanda) [inline assets/www/js/database.js:55]: Looks clearer, IMO. Shall add spacing though.

Comment (jdlrobson) [inline assets/www/js/database.js:55]: I still don't understand but ok :-)

**Comment by jdlrobson** [comment]
```
var api = new Api();
var listener = api.addEventListener( 'onprogresschange', callback ); ?
api.removeEventListener( listener);
```

?

**Comment by jdlrobson** [comment]
Bump. Also needs a rebase.

**Comment by yuvipanda** [comment]
Added better specification of the onProgressChange event (options param)

**Comment by yuvipanda** [comment]
Rebased!

File: assets/www/js/api.js [hunk 3]
```diff
@@ -101,19 +109,13 @@ define(['jquery'], function() {
...
-		$( '#upload-progress-bar' ).empty();
-		$( '<div>' ).css( 'width', percentage + '%').
-			appendTo( '#upload-progress-bar' );
-	};
-
 	Api.prototype.startUpload = function( sourceUri, filename ) {
 		var d = $.Deferred();
 		var that = this;
-		that.reportProgress( 0 );
+		that.options.onProgressChanged( 0 );
```
Comment (jdlrobson) [inline assets/www/js/api.js:115]: If options passed in to function is {} then this will break

File: assets/www/js/api.js [hunk 4]
```diff
@@ -3,7 +3,15 @@
...
+	var defaultOptions = {
+		onProgressChange: function() {
+		}
+	};
+
+	function Api( url, options ) {
+		options = options || defaultOptions;
+
+		this.options = options;
 		this.url = url;
```
Comment (jdlrobson) [inline assets/www/js/api.js:15]: should add

  this.options.onProgressChanged = this.options.onProgressChanged || function(){};

File: assets/www/js/api.js [hunk 5]
```diff
@@ -183,11 +185,11 @@ define(['jquery'], function() {
 		var that = this;
 		function sendImage( token ) {
 			var progress = 70;
-			that.reportProgress( progress );
+			that.options.onProgressChanged( progress );
```
Comment (jdlrobson) [inline assets/www/js/api.js:188]: cache variable?

File: assets/www/js/app.js [hunk 6]
```diff
@@ -19,11 +19,17 @@ function handleOpenURL(url)
...
+	function( $, l10n, geo, Api, templates, MonumentsApi, Monument, prefs, db, AdminTreeApi, Photo ) {
 
-	var api = new Api( WLMConfig.WIKI_API );
+	var api = new Api( WLMConfig.WIKI_API, {
+		onProgressChange: function( percent ) {
+				$( '#upload-progress-bar' ).empty();
+				$( '<div>' ).css( 'width', percent + '%').
+				appendTo( '#upload-progress-bar' );
+		}
+	});
```
Comment (jdlrobson) [inline assets/www/js/app.js:32]: whitespace after } needed

File: assets/www/js/app.js [hunk 7]
```diff
@@ -340,21 +346,28 @@ require( [ 'jquery', 'l10n', 'geo', 'api', 'templates', 'monuments', 'monument',
...
-					$( '#upload-latest-page .share' ).html( mw.msg( 'upload-latest-view' ) );
-					$( '#upload-latest-page .share a' ).attr( 'href', imageinfo.descriptionurl );
-
-					db.addUpload( curMonument, api.userName, fileUrl, imageinfo.url, fileName, true );
-					goBack(); // undo back button to skip upload progress page
-					goBack(); // undo back button to skip upload form
-					showPage( 'upload-latest-page' );
-				});
-			}).fail( function( data ) {
+			var photo = new Photo({
```
Comment (jdlrobson) [inline assets/www/js/app.js:349]: whitespace before {

File: assets/www/js/app.js [hunk 8]
```diff
@@ -340,21 +346,28 @@ require( [ 'jquery', 'l10n', 'geo', 'api', 'templates', 'monuments', 'monument',
...
+				contentURL: fileUrl,
+				fileTitle: fileName,
+				fileContent: text
+			});
```
Comment (jdlrobson) [inline assets/www/js/app.js:353]: whitespace

File: assets/www/js/app.js [hunk 9]
```diff
@@ -594,9 +607,10 @@ require( [ 'jquery', 'l10n', 'geo', 'api', 'templates', 'monuments', 'monument',
 				var uploadCompleteTemplate = templates.getTemplate( 'upload-completed-item-detail-template' );
 				$.each( uploads, function( i, upload ) {
 					var monument = JSON.parse( upload.monument );
```
Comment (jdlrobson) [inline assets/www/js/app.js:609]: don't like this... we should parse once at the top of the stack (in requestUploadsForUser)

File: assets/www/js/photo.js [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+		var d = $.Deferred();
+		d.notify( 'starting' );
+		mwApi.startUpload( data.contentURL, data.fileTitle ).done( function( fileKey, token ) {
+			d.notify( 'in-progress' );
+			mwApi.finishUpload( fileKey, data.fileTitle, comment, data.fileContent, token ).done( function( imageinfo ) {
+
+				d.resolve( imageinfo );
+			} ).fail( function() { d.reject( arguments ); } );
+
+		} ).fail( function() { d.reject( arguments ); } );
```
Comment (jdlrobson) [inline assets/www/js/photo.js:31]: don't do on one single line

File: assets/www/js/photo.js [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+	Photo.prototype.uploadTo = function( mwApi, comment ) {
+		var data = this.data;
+		var d = $.Deferred();
+		d.notify( 'starting' );
+		mwApi.startUpload( data.contentURL, data.fileTitle ).done( function( fileKey, token ) {
+			d.notify( 'in-progress' );
+			mwApi.finishUpload( fileKey, data.fileTitle, comment, data.fileContent, token ).done( function( imageinfo ) {
+
+				d.resolve( imageinfo );
+			} ).fail( function() { d.reject( arguments ); } );
```
Comment (jdlrobson) [inline assets/www/js/photo.js:29]: see below comment

**Comment by bvibber** [comment]
Hmm, I get this error when clicking the 'Upload' button:

08-13 13:11:18.972: E/Web Console(2953): Uncaught TypeError: Object #<Object> has no method 'onProgressChanged' at file:///android_asset/www/js/api.js:115

**Comment by yuvipanda** [comment]
@brion: Fixed. Was a rebase related snafu(I guess?)

**Comment by jdlrobson** [comment]
In process of merging with some adjustments

**Comment by jdlrobson** [comment]
Merged

**Pull request merged**